File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/BeneficiaryApplicationFragment.java
package org.mifos.mobilebanking.ui.fragments;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.widget.NestedScrollView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.Spinner;

import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.models.beneficiary.Beneficiary;
import org.mifos.mobilebanking.models.beneficiary.BeneficiaryPayload;
import org.mifos.mobilebanking.models.beneficiary.BeneficiaryUpdatePayload;
import org.mifos.mobilebanking.models.templates.beneficiary.AccountTypeOption;
import org.mifos.mobilebanking.models.templates.beneficiary.BeneficiaryTemplate;
import org.mifos.mobilebanking.presenters.BeneficiaryApplicationPresenter;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.enums.BeneficiaryState;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.BeneficiaryApplicationView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.Network;
import org.mifos.mobilebanking.utils.Toaster;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by dilpreet on 16/6/17.
 */

public class BeneficiaryApplicationFragment extends BaseFragment implements
        BeneficiaryApplicationView, AdapterView.OnItemSelectedListener {

    @BindView(R.id.sp_account_type)
    Spinner spAccountType;

    @BindView(R.id.ll_application_beneficiary)
    LinearLayout llApplicationBeneficiary;

    @BindView(R.id.til_account_number)
    TextInputLayout tilAccountNumber;

    @BindView(R.id.til_office_name)
    TextInputLayout tilOfficeName;

    @BindView(R.id.til_transfer_limit)
    TextInputLayout tilTransferLimit;

    @BindView(R.id.til_beneficiary_name)
    TextInputLayout tilBeneficiaryName;

    @BindView(R.id.layout_error)
    View layoutError;

    @BindView(R.id.view_flipper)
    NestedScrollView nsvBeneficiary;

    @Inject
    BeneficiaryApplicationPresenter presenter;

    private List<String> listAccountType = new ArrayList<>();
    private ArrayAdapter<String> accountTypeAdapter;

    private BeneficiaryState beneficiaryState;
    private Beneficiary beneficiary;
    private BeneficiaryTemplate beneficiaryTemplate;
    private int accountTypeId = -1;
    private View rootView;
    private SweetUIErrorHandler sweetUIErrorHandler;

    public static BeneficiaryApplicationFragment newInstance(BeneficiaryState beneficiaryState,
                                                             @Nullable Beneficiary beneficiary) {
        BeneficiaryApplicationFragment fragment = new BeneficiaryApplicationFragment();
        Bundle args = new Bundle();
        args.putSerializable(Constants.BENEFICIARY_STATE, beneficiaryState);
        if (beneficiary != null) {
            args.putParcelable(Constants.BENEFICIARY, beneficiary);
        }
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setToolbarTitle(getString(R.string.add_beneficiary));
        if (getArguments() != null) {
            beneficiaryState = (BeneficiaryState) getArguments()
                    .getSerializable(Constants.BENEFICIARY_STATE);
            if (beneficiaryState == BeneficiaryState.UPDATE) {
                beneficiary = getArguments().getParcelable(Constants.BENEFICIARY);
                setToolbarTitle(getString(R.string.update_beneficiary));
            } else if (beneficiaryState == BeneficiaryState.CREATE_QR) {
                beneficiary = getArguments().getParcelable(Constants.BENEFICIARY);
                setToolbarTitle(getString(R.string.add_beneficiary));
            } else {
                setToolbarTitle(getString(R.string.add_beneficiary));
            }
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_beneficiary_application, container, false);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        ButterKnife.bind(this, rootView);
        sweetUIErrorHandler = new SweetUIErrorHandler(getActivity(), rootView);
        showUserInterface();

        presenter.attachView(this);
        if (savedInstanceState == null) {
            presenter.loadBeneficiaryTemplate();
        }
        return rootView;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(Constants.TEMPLATE, beneficiaryTemplate);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (savedInstanceState != null) {
            showBeneficiaryTemplate((BeneficiaryTemplate) savedInstanceState.
                    getParcelable(Constants.TEMPLATE));
        }
    }

    /**
     * Setting up {@code accountTypeAdapter} and {@code} spAccountType
     */
    @Override
    public void showUserInterface() {
        accountTypeAdapter = new ArrayAdapter<>(getActivity(),
                android.R.layout.simple_spinner_item, listAccountType);
        accountTypeAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spAccountType.setOnItemSelectedListener(this);
        if (!Network.isConnected(getContext())) {
            spAccountType.setEnabled(false);
        }
        spAccountType.setAdapter(accountTypeAdapter);
    }

    /**
     * Fetches {@link BeneficiaryTemplate} from server and further updates the UI according to
     * {@link BeneficiaryState} which is initialized in {@code newInstance()} as
     * {@code beneficiaryState}
     * @param beneficiaryTemplate {@link BeneficiaryTemplate} fetched from server
     */
    @Override
    public void showBeneficiaryTemplate(BeneficiaryTemplate beneficiaryTemplate) {
        this.beneficiaryTemplate = beneficiaryTemplate;
        for (AccountTypeOption accountTypeOption : beneficiaryTemplate.getAccountTypeOptions()) {
            listAccountType.add(accountTypeOption.getValue());
        }
        accountTypeAdapter.notifyDataSetChanged();

        if (beneficiaryState == BeneficiaryState.UPDATE) {
            spAccountType.setSelection(accountTypeAdapter.getPosition(beneficiary
                    .getAccountType().getValue()));
            spAccountType.setEnabled(false);
            tilAccountNumber.getEditText().setText(beneficiary.getAccountNumber());
            tilAccountNumber.setEnabled(false);
            tilOfficeName.getEditText().setText(beneficiary.getOfficeName());
            tilOfficeName.setEnabled(false);
            tilBeneficiaryName.getEditText().setText(beneficiary.getName());
            tilTransferLimit.getEditText().setText(String.valueOf(beneficiary.getTransferLimit()));
        } else if (beneficiaryState == BeneficiaryState.CREATE_QR) {
            spAccountType.setSelection(beneficiary.getAccountType().getId());
            tilAccountNumber.getEditText().setText(beneficiary.getAccountNumber());
            tilOfficeName.getEditText().setText(beneficiary.getOfficeName());
        }
    }

    /**
     * Used for submitting a new or updating beneficiary application
     */
    @OnClick(R.id.btn_beneficiary_submit)
    public void submitBeneficiary() {

        tilAccountNumber.setErrorEnabled(false);
        tilOfficeName.setErrorEnabled(false);
        tilTransferLimit.setErrorEnabled(false);
        tilBeneficiaryName.setErrorEnabled(false);

        if (accountTypeId == -1) {
            Toaster.show(rootView, getString(R.string.choose_account_type));
            return;
        } else if (tilAccountNumber.getEditText().getText().toString().trim().equals("")) {
            tilAccountNumber.setError(getString(R.string.enter_account_number));
            return;
        } else if (tilOfficeName.getEditText().getText().toString().trim().equals("")) {
            tilOfficeName.setError(getString(R.string.enter_office_name));
            return;
        } else if (tilTransferLimit.getEditText().getText().toString().equals("")) {
            tilTransferLimit.setError(getString(R.string.enter_transfer_limit));
            return;
        } else if (tilTransferLimit.getEditText().getText().toString().equals(".")) {
            tilTransferLimit.setError(getString(R.string.invalid_amount));
            return;
        } else if (tilTransferLimit.getEditText().getText().toString().matches("^0*")) {
            tilTransferLimit.setError(getString(R.string.amount_greater_than_zero));
            return;
        } else if (tilBeneficiaryName.getEditText().getText().toString().trim().equals("")) {
            tilBeneficiaryName.setError(getString(R.string.enter_beneficiary_name));
            return;
        }

        if (beneficiaryState == BeneficiaryState.CREATE_MANUAL ||
                beneficiaryState == BeneficiaryState.CREATE_QR ) {
            submitNewBeneficiaryApplication();
        } else if (beneficiaryState == BeneficiaryState.UPDATE ) {
            submitUpdateBeneficiaryApplication();
        }

    }

    @OnClick(R.id.btn_try_again)
    public void onRetry() {
        if (Network.isConnected(getContext())) {
            presenter.loadBeneficiaryTemplate();
            sweetUIErrorHandler.hideSweetErrorLayoutUI(nsvBeneficiary, layoutError);
        } else {
            Toaster.show(rootView, getString(R.string.internet_not_connected));
        }
    }

    /**
     * Submit a new Beneficiary application
     */
    private void submitNewBeneficiaryApplication() {
        BeneficiaryPayload beneficiaryPayload = new BeneficiaryPayload();
        beneficiaryPayload.setAccountNumber(tilAccountNumber.getEditText().getText().toString());
        beneficiaryPayload.setOfficeName(tilOfficeName.getEditText().getText().toString());
        beneficiaryPayload.setAccountType(accountTypeId);
        beneficiaryPayload.setName(tilBeneficiaryName.getEditText().getText().toString());
        beneficiaryPayload.setTransferLimit(Integer.parseInt(tilTransferLimit.getEditText().
                getText().toString()));
        presenter.createBeneficiary(beneficiaryPayload);
    }

    /**
     * Updates an existing beneficiary application
     */
    private void submitUpdateBeneficiaryApplication() {
        BeneficiaryUpdatePayload payload = new BeneficiaryUpdatePayload();
        payload.setName(tilBeneficiaryName.getEditText().getText().toString());
        payload.setTransferLimit(Float.parseFloat(tilTransferLimit.getEditText().getText().
                toString()));
        presenter.updateBeneficiary(beneficiary.getId(), payload);
    }

    /**
     * Displays a {@link android.support.design.widget.Snackbar} on successfully creation of
     * Beneficiary and pops fragments in order to go back to {@link BeneficiaryListFragment}
     */
    @Override
    public void showBeneficiaryCreatedSuccessfully() {
        Toaster.show(rootView, getString(R.string.beneficiary_created_successfully));
        getActivity().finish();
    }

    /**
     * Displays a {@link android.support.design.widget.Snackbar} on successfully updation of
     * Beneficiary and pops fragments in order to go back to {@link BeneficiaryListFragment}
     */
    @Override
    public void showBeneficiaryUpdatedSuccessfully() {
        Toaster.show(rootView, getString(R.string.beneficiary_updated_successfully));
        getActivity().getSupportFragmentManager().popBackStack();
        getActivity().getSupportFragmentManager().popBackStack();

    }


    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        accountTypeId = beneficiaryTemplate.getAccountTypeOptions().get(position).getId();
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }

    /**
     * It is called whenever any error occurs while executing a request
     * @param msg Error message that tells the user about the problem.
     */
    @Override
    public void showError(String msg ) {
        if (!Network.isConnected(getContext())) {
            sweetUIErrorHandler.showSweetNoInternetUI(nsvBeneficiary, layoutError);
        } else {
            sweetUIErrorHandler.showSweetErrorUI(msg, nsvBeneficiary, layoutError);
            Toaster.show(rootView, msg);
        }
    }

    @Override
    public void setVisibility(int state) {
        llApplicationBeneficiary.setVisibility(state);
    }

    @Override
    public void showProgress() {
        showProgressBar();
    }

    @Override
    public void hideProgress() {
        hideProgressBar();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        hideProgress();
        presenter.detachView();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/BeneficiaryApplicationFragment.kt
package org.mifos.mobile.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentBeneficiaryApplicationBinding
import org.mifos.mobile.models.beneficiary.Beneficiary
import org.mifos.mobile.models.beneficiary.BeneficiaryPayload
import org.mifos.mobile.models.beneficiary.BeneficiaryUpdatePayload
import org.mifos.mobile.models.templates.beneficiary.BeneficiaryTemplate
import org.mifos.mobile.ui.activities.base.BaseActivity
import org.mifos.mobile.ui.enums.BeneficiaryState
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.BeneficiaryUiState
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.Network
import org.mifos.mobile.utils.ParcelableAndSerializableUtils.getCheckedParcelable
import org.mifos.mobile.utils.ParcelableAndSerializableUtils.getCheckedSerializable
import org.mifos.mobile.utils.Toaster
import org.mifos.mobile.viewModels.BeneficiaryApplicationViewModel

/**
 * Created by dilpreet on 16/6/17.
 */
@AndroidEntryPoint
class BeneficiaryApplicationFragment : BaseFragment() {

    private var _binding: FragmentBeneficiaryApplicationBinding? = null
    private val binding get() = _binding!!

    private val viewModel: BeneficiaryApplicationViewModel by viewModels()

    private val listAccountType: MutableList<String?> = ArrayList()
    private var beneficiaryState: BeneficiaryState? = null
    private var beneficiary: Beneficiary? = null
    private var beneficiaryTemplate: BeneficiaryTemplate? = null
    private var accountTypeId: Int? = -1
    private var sweetUIErrorHandler: SweetUIErrorHandler? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setToolbarTitle(getString(R.string.add_beneficiary))
        if (arguments != null) {
            beneficiaryState = requireArguments().getCheckedSerializable(
                    BeneficiaryState::class.java,
                    Constants.BENEFICIARY_STATE
                ) as BeneficiaryState
            when (beneficiaryState) {
                BeneficiaryState.UPDATE -> {
                    beneficiary = arguments?.getCheckedParcelable(
                        Beneficiary::class.java,
                        Constants.BENEFICIARY
                    )
                    setToolbarTitle(getString(R.string.update_beneficiary))
                }

                BeneficiaryState.CREATE_QR -> {
                    beneficiary = arguments?.getCheckedParcelable(
                        Beneficiary::class.java,
                        Constants.BENEFICIARY
                    )
                    setToolbarTitle(getString(R.string.add_beneficiary))
                }

                else -> {
                    setToolbarTitle(getString(R.string.add_beneficiary))
                }
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentBeneficiaryApplicationBinding.inflate(inflater, container, false)
        sweetUIErrorHandler = SweetUIErrorHandler(activity, binding.root)
        showUserInterface()
        if (savedInstanceState == null) {
            viewModel.loadBeneficiaryTemplate()
        }
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.beneficiaryUiState.collect {
                    when (it) {
                        is BeneficiaryUiState.Loading -> showProgress()
                        is BeneficiaryUiState.ShowError -> {
                            hideProgress()
                            showError(getString(it.message))
                        }

                        is BeneficiaryUiState.SetVisibility -> {
                            hideProgress()
                            setVisibility(it.visibility)
                        }

                        is BeneficiaryUiState.ShowBeneficiaryTemplate -> {
                            hideProgress()
                            showBeneficiaryTemplate(it.beneficiaryTemplate)
                        }

                        is BeneficiaryUiState.CreatedSuccessfully -> {
                            hideProgress()
                            showBeneficiaryCreatedSuccessfully()
                        }

                        is BeneficiaryUiState.UpdatedSuccessfully -> {
                            hideProgress()
                            showBeneficiaryUpdatedSuccessfully()
                        }

                        is BeneficiaryUiState.Initial -> {}

                        else -> throw IllegalStateException("Undesired $it")
                    }
                }
            }
        }

        with(binding) {
            btnBeneficiarySubmit.setOnClickListener {
                submitBeneficiary()
            }
            layoutError.btnTryAgain.setOnClickListener {
                onRetry()
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(Constants.TEMPLATE, beneficiaryTemplate)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState != null) {
            showBeneficiaryTemplate(
                savedInstanceState.getCheckedParcelable(
                    BeneficiaryTemplate::class.java,
                    Constants.TEMPLATE
                )
            )
        }
    }

    /**
     * Setting up `accountTypeAdapter` and `` spAccountType
     */
    fun showUserInterface() {
        with(binding) {
            accountTypeField.setOnItemClickListener { _, _, position, _ ->
                accountTypeId = beneficiaryTemplate?.accountTypeOptions?.get(position)?.id
            }
            accountTypeFieldParent.isEnabled = Network.isConnected(context)
            accountTypeField.setSimpleItems(listAccountType.toTypedArray())
        }
    }

    /**
     * Fetches [BeneficiaryTemplate] from server and further updates the UI according to
     * [BeneficiaryState] which is initialized in `newInstance()` as
     * `beneficiaryState`
     *
     * @param beneficiaryTemplate [BeneficiaryTemplate] fetched from server
     */
    private fun showBeneficiaryTemplate(beneficiaryTemplate: BeneficiaryTemplate?) {
        this.beneficiaryTemplate = beneficiaryTemplate
        for ((_, _, value) in beneficiaryTemplate?.accountTypeOptions!!) {
            listAccountType.add(value)
        }
        binding.accountTypeField.setSimpleItems(listAccountType.toTypedArray())
        if (beneficiaryState == BeneficiaryState.UPDATE) {
            with(binding) {
                accountTypeField.setText(beneficiary?.accountType?.value!!, false)
                accountTypeFieldParent.isEnabled = false
                tilAccountNumber.editText?.setText(beneficiary?.accountNumber)
                tilAccountNumber.isEnabled = false
                tilOfficeName.editText?.setText(beneficiary?.officeName)
                tilOfficeName.isEnabled = false
                tilBeneficiaryName.editText?.setText(beneficiary?.name)
                tilTransferLimit.editText?.setText(beneficiary?.transferLimit.toString())
            }
        } else if (beneficiaryState == BeneficiaryState.CREATE_QR) {
            with(binding) {
                accountTypeField.setText(beneficiary?.accountType?.value!!, false)
                tilAccountNumber.editText?.setText(beneficiary?.accountNumber)
                tilOfficeName.editText?.setText(beneficiary?.officeName)
            }
        }
    }

    /**
     * Used for submitting a new or updating beneficiary application
     */
    private fun submitBeneficiary() {
        with(binding) {
            tilAccountNumber.isErrorEnabled = false
            tilOfficeName.isErrorEnabled = false
            tilTransferLimit.isErrorEnabled = false
            tilBeneficiaryName.isErrorEnabled = false
        }
        with(binding) {
            when {
                accountTypeId == -1 -> {
                    Toaster.show(root, getString(R.string.choose_account_type))
                    return
                }

                tilAccountNumber.editText?.text.isNullOrBlank() -> {
                    tilAccountNumber.error = getString(R.string.enter_account_number)
                    return
                }

                tilOfficeName.editText?.text.isNullOrBlank() -> {
                    tilOfficeName.error = getString(R.string.enter_office_name)
                    return
                }

                tilTransferLimit.editText?.text.toString().isEmpty() -> {
                    tilTransferLimit.error = getString(R.string.enter_transfer_limit)
                    return
                }

                tilTransferLimit.editText?.text.toString() == "." -> {
                    tilTransferLimit.error = getString(R.string.invalid_amount)
                    return
                }

                tilTransferLimit.editText?.text.toString().matches("^0*".toRegex()) -> {
                    tilTransferLimit.error = getString(R.string.amount_greater_than_zero)
                    return
                }

                tilBeneficiaryName.editText?.text.isNullOrBlank() -> {
                    tilBeneficiaryName.error = getString(R.string.enter_beneficiary_name)
                    return
                }

                else -> {}
            }
        }

        if (beneficiaryState == BeneficiaryState.CREATE_MANUAL ||
            beneficiaryState == BeneficiaryState.CREATE_QR
        ) {
            submitNewBeneficiaryApplication()
        } else if (beneficiaryState == BeneficiaryState.UPDATE) {
            submitUpdateBeneficiaryApplication()
        }
    }

    private fun onRetry() {
        if (Network.isConnected(context)) {
            viewModel.loadBeneficiaryTemplate()
            sweetUIErrorHandler?.hideSweetErrorLayoutUI(
                binding.viewFlipper,
                binding.layoutError.root,
            )
        } else {
            Toaster.show(binding.root, getString(R.string.internet_not_connected))
        }
    }

    /**
     * Submit a new Beneficiary application
     */
    private fun submitNewBeneficiaryApplication() {
        val beneficiaryPayload = BeneficiaryPayload()
        with(binding) {
            beneficiaryPayload.accountNumber = tilAccountNumber.editText?.text.toString()
            beneficiaryPayload.officeName = tilOfficeName.editText?.text.toString()
            beneficiaryPayload.name = tilBeneficiaryName.editText?.text.toString()
            beneficiaryPayload.transferLimit =
                tilTransferLimit.editText?.text.toString().toInt().toFloat()
        }

        beneficiaryPayload.accountType = accountTypeId
        viewModel.createBeneficiary(beneficiaryPayload)
    }

    /**
     * Updates an existing beneficiary application
     */
    private fun submitUpdateBeneficiaryApplication() {
        val payload = BeneficiaryUpdatePayload()
        payload.name = binding.tilBeneficiaryName.editText?.text.toString()
        payload.transferLimit = binding.tilTransferLimit.editText?.text.toString().toFloat()
        viewModel.updateBeneficiary(beneficiary?.id?.toLong(), payload)
    }

    /**
     * Displays a {@link Snackbar} on successfully creation of
     * Beneficiary and pops fragments in order to go back to [BeneficiaryListFragment]
     */
    private fun showBeneficiaryCreatedSuccessfully() {
        Toaster.show(binding.tilTransferLimit, getString(R.string.beneficiary_created_successfully))
        activity?.finish()
    }

    /**
     * Displays a {@link Snackbar} on successfully updation of
     * Beneficiary and pops fragments in order to go back to [BeneficiaryListFragment]
     */
    private fun showBeneficiaryUpdatedSuccessfully() {
        Toaster.show(binding.root, getString(R.string.beneficiary_updated_successfully))
        activity?.supportFragmentManager?.popBackStack()
        activity?.supportFragmentManager?.popBackStack()
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param msg Error message that tells the user about the problem.
     */
    fun showError(msg: String?) {
        if (!Network.isConnected(context)) {
            sweetUIErrorHandler?.showSweetNoInternetUI(
                binding.viewFlipper,
                binding.layoutError.root,
            )
        } else {
            sweetUIErrorHandler?.showSweetErrorUI(
                msg,
                binding.viewFlipper,
                binding.layoutError.root,
            )
            Toaster.show(binding.root, msg)
        }
    }

    private fun setVisibility(state: Int) {
        binding.llApplicationBeneficiary.visibility = state
    }

    override fun onResume() {
        super.onResume()
        (activity as? BaseActivity)?.showToolbar()
    }

    fun showProgress() {
        showProgressBar()
    }

    fun hideProgress() {
        hideProgressBar()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        hideProgress()
        _binding = null
    }

    companion object {
        fun newInstance(
            beneficiaryState: BeneficiaryState?,
            beneficiary: Beneficiary?,
        ): BeneficiaryApplicationFragment {
            val fragment = BeneficiaryApplicationFragment()
            val args = Bundle()
            args.putSerializable(Constants.BENEFICIARY_STATE, beneficiaryState)
            if (beneficiary != null) {
                args.putParcelable(Constants.BENEFICIARY, beneficiary)
            }
            fragment.arguments = args
            return fragment
        }
    }
}
